python_modules/libraries/dagster-azure/dagster_azure/adls2/object_store.py

```diff
@@ -0,0 +1,155 @@
+import logging
+import re
+import sys
+from io import BytesIO, StringIO
+
+from azure.core.exceptions import ResourceNotFoundError
+
+from dagster import check
+from dagster.core.definitions.events import ObjectStoreOperation, ObjectStoreOperationType
+from dagster.core.storage.object_store import ObjectStore
+from dagster.core.types.marshal import SerializationStrategy
+from dagster.seven import urlparse
+
+from dagster_azure.blob.utils import create_blob_client
+
+DEFAULT_LEASE_DURATION = 60  # One minute
+
+
+class ADLS2ObjectStore(ObjectStore):
+    def __init__(self, file_system, client, lease_duration=DEFAULT_LEASE_DURATION):
+        self.adls2_client = client
+        self.file_system_client = self.adls2_client.get_file_system_client(file_system)
+        # We also need a blob client to handle copying as ADLS doesn't have a copy API yet
+        self.blob_client = create_blob_client(
+            client.account_name,
+            # client.credential is non-null if a secret key was used to authenticate
+            client.credential.account_key if client.credential is not None
+            # otherwise the SAS token will be in the query string of the URL
+            else urlparse(client.url).query,
+        )
+        self.blob_container_client = self.blob_client.get_container_client(file_system)
+
+        self.lease_duration = lease_duration
+        self.file_system_client.get_file_system_properties()
+        super(ADLS2ObjectStore, self).__init__('adls2', sep='/')
+
+    def set_object(self, key, obj, serialization_strategy=None):
+        check.str_param(key, 'key')
+
+        logging.info('Writing ADLS2 object at: ' + self.uri_for_key(key))
+
+        # cannot check obj since could be arbitrary Python object
+        check.inst_param(
+            serialization_strategy, 'serialization_strategy', SerializationStrategy
+        )  # cannot be none here
+
+        if self.has_object(key):
+            logging.warning('Removing existing ADLS2 key: {key}'.format(key=key))
+            self.rm_object(key)
+
+        file = self.file_system_client.create_file(key)
+        with file.acquire_lease(self.lease_duration) as lease:
+            with BytesIO() as bytes_io:
```

Comment:
Since this is used in s3/object_store, adls2/object_store, and maybe blob/object_store as well, I think it's worth factoring out into a shared helper.